flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TriggerManager.java

```diff
@@ -0,0 +1,296 @@
+    LOG.info("Initializing state restored: {}", context.isRestored());
+    this.lock = lockFactory.createLock();
+    this.recoveryLock = lockFactory.createRecoveryLock();
+    if (context.isRestored()) {
+      isCleanUp = true;
+    }
+  }
+
+  @Override
+  public void processElement(TableChange change, Context ctx, Collector<Trigger> out)
+      throws Exception {
+    init(out, ctx.timerService());
+
+    long current = ctx.timerService().currentProcessingTime();
+    Long nextTime = nextEvaluationTime.value();
+
+    // Add the new changes to the already accumulated ones
+    List<TableChange> accumulated = Lists.newArrayList(accumulatedChanges.get());
+    accumulated.forEach(tableChange -> tableChange.merge(change));
+    accumulatedChanges.update(accumulated);
+
+    if (nextTime == null) {
+      checkAndFire(current, ctx.timerService(), out);
+    } else {
+      LOG.info(
+          "Trigger manager rate limiter triggered current: {}, next: {}, accumulated changes: {}",
+          current,
+          nextTime,
+          accumulated);
+      rateLimiterTriggeredCounter.inc();
+    }
+  }
+
+  @Override
+  public void onTimer(long timestamp, OnTimerContext ctx, Collector<Trigger> out) throws Exception {
+    init(out, ctx.timerService());
+    nextEvaluationTime.clear();
+    checkAndFire(ctx.timerService().currentProcessingTime(), ctx.timerService(), out);
+  }
+
+  private void checkAndFire(long current, TimerService timerService, Collector<Trigger> out)
+      throws Exception {
+    if (isCleanUp) {
+      if (recoveryLock.isHeld()) {
+        LOG.debug("The cleanup lock is still held at {}", current);
+        schedule(timerService, current + lockCheckDelayMs);
+        return;
+      } else {
+        LOG.info("The cleanup is finished at {}", current);
+        isCleanUp = false;
+      }
+    }
+
+    List<TableChange> changes = Lists.newArrayList(accumulatedChanges.get());
+    List<Long> times = Lists.newArrayList(lastTriggerTimes.get());
+    Integer taskToStart = nextTrigger(evaluators, changes, times, current, startsFrom);
+    if (taskToStart == null) {
+      // Nothing to execute
+      if (startsFrom == 0) {
+        nothingToTriggerCounter.inc();
+        LOG.debug("Nothing to execute at {} for collected: {}", current, changes);
+      }
+
+      startsFrom = 0;
+      return;
+    }
+
+    if (lock.tryLock()) {
+      TableChange change = changes.get(taskToStart);
+      SerializableTable table =
+          (SerializableTable) SerializableTable.copyOf(tableLoader.loadTable());
+      out.collect(Trigger.create(current, table, taskToStart));
+      LOG.debug("Fired event with time: {}, collected: {} for {}", current, change, table.name());
+      triggerCounters.get(taskToStart).inc();
+      changes.set(taskToStart, TableChange.empty());
+      accumulatedChanges.update(changes);
+      schedule(timerService, current + minFireDelayMs);
+      startsFrom = taskToStart + 1;
```

Comment:
might be a bit more readable to do the `mod` here (than in the `nextTrigger` method)

Reply:
This is needed, so we can differentiate the 2 situations
- We start now
- We reached the last element, and we need to start from the beginning

Reply:
where do we need to differentiate the 2 situations?

Reply:
I was saying perform the mod operation here to make sure `startsFrom` is always a valid position. This way, if we log this value, it will show correct value.

```
startsFrom = (taskToStart + 1) % evaluators.size();
```

instead of in `nextTrigger` method
```
int normalizedStartingPos = startPos % evaluators.size();
```

This should affect fairness, right?

Reply:
If the `startsFrom` is 0, then we did not execute anything in this loop.
If the `startsFrom` is anything but 0, it means we executed something. If we would execute the last task, and we calculate the `startsFrom` as `(startsFrom + 1) % size` then it would be 0.

```
      // Nothing to execute
      if (startsFrom == 0) {
        nothingToTriggerCounter.inc();
        LOG.debug("Nothing to execute at {} for collected: {}", current, accumulatedChanges);
      }
```